src/kv/LockResolver.cc

```diff
@@ -253,11 +271,179 @@ void LockResolver::resolvePessimisticLock(Backoffer & bo, LockPtr lock, std::uno
     }
 }
 
+void LockResolver::resolveLockAsync(Backoffer & bo, LockPtr lock, TxnStatus & status)
+{
+    AsyncResolveDataPtr resolve_data{};
+    resolve_data = checkAllSecondaries(bo, lock, status);
+
+    status.commit_ts = resolve_data->commit_ts;
+
+    resolve_data->keys.push_back(lock->primary);
+    auto [keys_by_region, region_id] = cluster->region_cache->groupKeysByRegion(bo, resolve_data->keys);
+    (void)region_id;
+    std::vector<std::thread> threads;
+    std::atomic<int> errors{};
+    for (auto & pair : keys_by_region)
+    {
+        auto & region_id = pair.first;
+        auto & locks = pair.second;
+        threads.emplace_back([&]() {
+            try
+            {
+                resolveRegionLocks(bo, lock, region_id, locks, status);
+            }
+            catch (Exception & e)
+            {
+                errors.fetch_add(1);
+                log->error("resolveRegionLocks error: " + e.displayText());
+            }
+        });
+    }
+
+    for (auto & t : threads)
+    {
+        t.join();
+    }
+
+    if (errors.load() > 0)
+    {
+        throw Exception("AsyncCommit recovery finished with errors", ErrorCodes::UnknownError);
+    }
+}
+
+void LockResolver::resolveRegionLocks(
+    Backoffer & bo, LockPtr lock, RegionVerID region_id, std::vector<std::string> & keys, TxnStatus & status)
+{
+    auto req = std::make_shared<::kvrpcpb::ResolveLockRequest>();
+    req->set_start_version(lock->txn_id);
+
+    if (status.isCommited())
+    {
+        req->set_commit_version(status.commit_ts);
+    }
+
+    for (auto & key : keys)
+    {
+        auto k = req->add_keys();
+        *k = key;
+    }
+
+    RegionClient client(cluster, region_id);
+    std::shared_ptr<::kvrpcpb::ResolveLockResponse> response{};
+
+    try
+    {
+        response = client.sendReqToRegion(bo, req);
+    }
+    catch (Exception & e)
+    {
+        bo.backoff(boRegionMiss, e);
+
+        auto [regions, region_id] = cluster->region_cache->groupKeysByRegion(bo, keys);
```

Comment:
it seems we should drop the miss regions from cache before retry ?

Reply:
```suggestion
        cluster->region_cache->dropRegion(region_id);
```
Like this?

Reply:
not really, I not sure if the sendRequest function has dropped the problematic region. Could you please check it out?

Reply:
It seems the cache is refreshed before the exception is caught. So this is no problem for now.